fw/sock_clnt.c

```diff
@@ -620,28 +624,90 @@ tfw_sock_clnt_cfgend(void)
 	return 0;
 }
 
+static int
+tfw_listen_socks_array_cmp(const void *l, const void *r)
+{
+	TfwAddr *a = &(*(TfwListenSock **)l)->addr;
+	TfwAddr *b = &(*(TfwListenSock **)r)->addr;
+	int cmp = memcmp(&a->sin6_addr, &b->sin6_addr, sizeof(a->sin6_addr));
+
+	if (cmp)
+		return cmp;
+	else
+		return (a->sin6_port < b->sin6_port) ? -1 :
+	                                           (a->sin6_port > b->sin6_port);
+}
+
 /**
  * Start listening on all existing sockets (added via "listen" configuration
  * entries).
  */
 static int
 tfw_sock_clnt_start(void)
 {
-	int r;
+	int r = 0;
 	TfwListenSock *ls;
+	TfwListenSock *ls_reconf;
+	TfwListenSock *tmp;
+	size_t i, listen_socks_sz = tfw_listen_socks_sz;
+	TfwListenSock **listen_socks_array = kmalloc(tfw_listen_socks_sz *
+		                                         sizeof(listen_socks_array[0]),
+		                                         GFP_KERNEL);
+	TfwListenSock **ls_found;
+	bool *toched = kzalloc(tfw_listen_socks_sz, GFP_KERNEL);
+
+	i = 0;
+	list_for_each_entry(ls, &tfw_listen_socks, list)
+		listen_socks_array[i++] = ls;
+	sort(listen_socks_array, tfw_listen_socks_sz,
+		 sizeof(listen_socks_array[0]), tfw_listen_socks_array_cmp, NULL);
+
+	list_for_each_entry_safe(ls_reconf, tmp, &tfw_listen_socks_reconf, list) {
+		if (!(ls_found = bsearch(&ls_reconf, listen_socks_array,
+			                     tfw_listen_socks_sz,
+			                     sizeof(listen_socks_array[0]),
+			                     tfw_listen_socks_array_cmp))) {
+				list_del(&ls_reconf->list);
+				list_add(&ls_reconf->list, &tfw_listen_socks);
+
+				if ((r = tfw_listen_sock_start(ls_reconf))) {
+					T_ERR_ADDR("can't start listening on", &ls_reconf->addr,
+						   TFW_WITH_PORT);
+					goto err;
+				}
+				tfw_classifier_add_inport(tfw_addr_port(&ls_reconf->addr));
+				listen_socks_sz++;
+		} else {
+			toched[ls_found - &listen_socks_array[0]] = true;
+		}
+	}
 
-	if (tfw_runstate_is_reconfig())
-		return 0;
+	for (i = 0; i < tfw_listen_socks_sz; ++i) {
+		if (toched[i])
+			continue;
 
-	list_for_each_entry(ls, &tfw_listen_socks, list) {
-		if ((r = tfw_listen_sock_start(ls))) {
-			T_ERR_ADDR("can't start listening on", &ls->addr,
-				   TFW_WITH_PORT);
-			goto err;
-		}
+		ls = listen_socks_array[i];
+		list_del(&ls->list);
+		if (!ls->sk)
+			continue;
+		ss_release(ls->sk);
+		ls->sk = NULL;
+		kfree(ls);
+
+		tfw_classifier_remove_inport(tfw_addr_port(&ls->addr));
+		listen_socks_sz--;
 	}
 
-	return 0;
+	tfw_listen_socks_sz = listen_socks_sz;
+
+done:
+	kfree(listen_socks_array);
+
+	list_for_each_entry_safe(ls, tmp, &tfw_listen_socks_reconf, list)
+		kfree(ls);
```

Comment:
It seems this should be for error condition only. In normal circumstances I think we can assert that the list was drained either by found sockets or by the loop over `touched`, i.e. it's empty

Reply:
Oh, I see. Could you please add a comment here that we're freeing remaining touched sockets?